kaldb/src/main/java/com/slack/kaldb/server/RecoveryTaskFactory.java

```diff
@@ -0,0 +1,260 @@
+import java.util.stream.Collectors;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is responsible for the indexer startup operations like stale live snapshot cleanup.
+ * determining the start indexing offset from metadata and optionally creating a recovery task etc.
+ */
+public class RecoveryTaskFactory {
+  private static final Logger LOG = LoggerFactory.getLogger(RecoveryTaskFactory.class);
+  private static final int SNAPSHOT_OPERATION_TIMEOUT_SECS = 10;
+
+  private final SnapshotMetadataStore snapshotMetadataStore;
+  private final RecoveryTaskMetadataStore recoveryTaskMetadataStore;
+  private final String partitionId;
+  private final long maxOffsetDelay;
+
+  public static final String STALE_SNAPSHOT_DELETE_SUCCESS = "stale_snapshot_delete_success";
+  public static final String STALE_SNAPSHOT_DELETE_FAILED = "stale_snapshot_delete_failed";
+
+  private final Counter snapshotDeleteSuccess;
+  private final Counter snapshotDeleteFailed;
+
+  public RecoveryTaskFactory(
+      SnapshotMetadataStore snapshotMetadataStore,
+      RecoveryTaskMetadataStore recoveryTaskMetadataStore,
+      String partitionId,
+      long maxOffsetDelay,
+      MeterRegistry meterRegistry) {
+    checkState(
+        partitionId != null && !partitionId.isEmpty(), "partitionId shouldn't be null or empty");
+    checkState(maxOffsetDelay > 0, "maxOffsetDelay should be a positive number");
+    this.snapshotMetadataStore = snapshotMetadataStore;
+    this.recoveryTaskMetadataStore = recoveryTaskMetadataStore;
+    this.partitionId = partitionId;
+    this.maxOffsetDelay = maxOffsetDelay;
+
+    snapshotDeleteSuccess = meterRegistry.counter(STALE_SNAPSHOT_DELETE_SUCCESS);
+    snapshotDeleteFailed = meterRegistry.counter(STALE_SNAPSHOT_DELETE_FAILED);
+  }
+
+  @VisibleForTesting
+  public List<SnapshotMetadata> getStaleLiveSnapshots(List<SnapshotMetadata> snapshots) {
+    return snapshots
+        .stream()
+        .filter(snapshotMetadata -> snapshotMetadata.partitionId.equals(partitionId))
+        .filter(SnapshotMetadata::isLive)
+        .collect(Collectors.toUnmodifiableList());
+  }
+
+  @VisibleForTesting
+  public List<SnapshotMetadata> deleteStaleLiveSnapshots(List<SnapshotMetadata> snapshots) {
+    List<SnapshotMetadata> staleSnapshots = getStaleLiveSnapshots(snapshots);
+    LOG.info("Deleting {} stale snapshots: {}", staleSnapshots.size(), staleSnapshots);
+    int deletedSnapshotCount = deleteSnapshots(snapshotMetadataStore, staleSnapshots);
+
+    int failedDeletes = staleSnapshots.size() - deletedSnapshotCount;
+    if (failedDeletes > 0) {
+      LOG.warn("Failed to delete {} live snapshots", failedDeletes);
+      throw new IllegalStateException("Failed to delete stale live snapshots");
+    }
+
+    return staleSnapshots;
+  }
+
+  // Get the highest offset for which data is durable for a partition.
+  @VisibleForTesting
+  public long getHighestDurableOffsetForPartition(
```

Comment:
The PR description says this PR only cleans up old indexer stale state. Is this method needed for that? Or for the followup PR ?

Reply:
This method is used for creating the recovery task.

Reply:
Done.